numpy/ma/core.py

```diff
@@ -3248,10 +3234,22 @@ def __setitem__(self, indx, value):
                 _mask[indx] = mval
         elif not self._hardmask:
             # Unshare the mask if necessary to avoid propagation
+            # We want to remove the unshare logic from this place in the
+            # future. Note that _sharedmask has lots of false positives.
             if not self._isfield:
-                _oldsharedmask = self._oldsharedmask
+                if self._sharedmask and not (
+                        # No one else _mask, self._mask + 1 holds a reference
+                        # and the array owns its own data
+                        (sys.getrefcount(_mask) == 3) and _mask.flags.owndata):
```

Comment:
I'm sure this check is right, but I have trouble telling from the comment :-). What are the three refs that are allowed, and why is it safe to discard the `_sharedmask` in this case?